### PR TITLE
Fixed issue #6831

### PR DESCRIPTION
**Pull request by jkrshnmenon**
Implemented shellcode compilation and encoding.

### CODE REVIEW

File: libr/core/cmd_egg.c [hunk 1]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
+	if (!egg->bin){
+		egg->bin = r_buf_new ();
+	}
+	if (!(b = r_egg_get_bin (egg))) {
+		eprintf ("r_egg_get_bin: invalid egg :(\n");
+		goto fail;
+	}
+	r_egg_finalize (egg);
+	for (i = 0; i < b->length; i++){
+		printf ("%02x", b->buf[i]);
```
Comment (radare) [inline libr/core/cmd_egg.c]: use r_cons_printf instead of printf

File: libr/core/cmd_egg.c [hunk 2]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
+	}
+	printf ("\n");
```
Comment (radare) [inline libr/core/cmd_egg.c]: use r_cons_newline()

File: libr/core/cmd_egg.c [hunk 3]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
+	return 0;
+	fail:
```
Comment (radare) [inline libr/core/cmd_egg.c]: wat

File: libr/core/cmd_egg.c [hunk 4]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
 
+static int compileShellcode (REgg *egg, const char *input){
+	int i = 0;
+	RBuffer *b;
+	if (!r_egg_shellcode (egg, input)) {
+		eprintf ("Unknown shellcode '%s'\n", input);
+	}
+	if (!r_egg_assemble (egg)){
+		eprintf ("r_egg_assemble : invalid assembly\n");
+		goto fail;
```
Comment (radare) [inline libr/core/cmd_egg.c]: return here, you dont have anything to free

File: libr/core/cmd_egg.c [hunk 5]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
+	}
+	if (!egg->bin){
+		egg->bin = r_buf_new ();
+	}
+	if (!(b = r_egg_get_bin (egg))) {
```
Comment (radare) [inline libr/core/cmd_egg.c]: i would define RBuffer in here

File: libr/core/cmd_egg.c [hunk 6]
```diff
@@ -27,6 +27,33 @@ static void showBuffer(RBuffer *b) {
...
+		eprintf ("r_egg_get_bin: invalid egg :(\n");
+		goto fail;
+	}
+	r_egg_finalize (egg);
+	for (i = 0; i < b->length; i++){
```
Comment (radare) [inline libr/core/cmd_egg.c]: use r2-indent or add space before {

File: libr/core/cmd_egg.c [hunk 7]
```diff
@@ -107,13 +134,26 @@ static int cmd_egg(void *data, const char *input) {
 			eprintf ("Cannot compile\n");
 		break;
 	case 'p': // "gp"
-		cmd_egg_option (egg, "egg.padding", input);
+		if (*input && *(input+2)) {
+			r_egg_padding(egg, input+2);
```
Comment (radare) [inline libr/core/cmd_egg.c]: use r2-indent plz missing spaces around +

File: libr/core/cmd_egg.c [hunk 8]
```diff
@@ -107,13 +134,26 @@ static int cmd_egg(void *data, const char *input) {
 			eprintf ("Cannot compile\n");
 		break;
 	case 'p': // "gp"
-		cmd_egg_option (egg, "egg.padding", input);
+		if (*input && *(input+2)) {
```
Comment (radare) [inline libr/core/cmd_egg.c]: more missing spaces, and use [] instead of *

File: libr/core/cmd_egg.c [hunk 9]
```diff
@@ -107,13 +134,26 @@ static int cmd_egg(void *data, const char *input) {
...
+			r_egg_padding(egg, input+2);
+		}
+		//cmd_egg_option (egg, "egg.padding", input);
 		break;
 	case 'e': // "ge"
-		cmd_egg_option (egg, "egg.encoder", input);
+		if (*input && *(input+2)) {
```
Comment (radare) [inline libr/core/cmd_egg.c]: use [] instead of *

Comment (radare) [inline libr/core/cmd_egg.c]: and add missing spaces

File: libr/core/cmd_egg.c [hunk 10]
```diff
@@ -107,13 +134,26 @@ static int cmd_egg(void *data, const char *input) {
...
+			if (!r_egg_encode (egg,input+2)) {
+				eprintf ("Invalid encoder '%s'\n", input+2);
+			}
+		}
+		//cmd_egg_option (egg, "egg.encoder", input);
 		break;
 	case 'i': // "gi"
-		cmd_egg_option (egg, "egg.shellcode", input);
+		if (*input && *(input+2)) {
+			compileShellcode(egg,input+2);
```
Comment (radare) [inline libr/core/cmd_egg.c]: missing spaces before (, around +, use [] instead of *

File: libr/core/cmd_egg.c [hunk 11]
```diff
@@ -107,13 +134,26 @@ static int cmd_egg(void *data, const char *input) {
...
+		}
```
Comment (radare) [inline libr/core/cmd_egg.c]: else block goes in the same line as }, but if it's empty... why have it?

Comment (jkrshnmenon) [inline libr/core/cmd_egg.c]: Was planning on a nice error message. Something like "usage : gi <shellcode>". Is that a good idea ?

**Comment by Maijin** [comment]
@jkrshnmenon by the way, notice that I've gathered the Ragg2 issue in a META issue so it's easier to see what there is to do.

**Comment by Maijin** [comment]
https://github.com/radare/radare2/issues/6949

File: libr/core/cmd_egg.c [hunk 12]
```diff
@@ -107,13 +132,26 @@ static int cmd_egg(void *data, const char *input) {
...
+				eprintf ("Invalid encoder '%s'\n", input+2);
+			}
+		}
+		//cmd_egg_option (egg, "egg.encoder", input);
 		break;
 	case 'i': // "gi"
-		cmd_egg_option (egg, "egg.shellcode", input);
+		if (input[0] && input[2]) {
+			compileShellcode (egg,input + 2);
+		}
```
Comment (radare) [inline libr/core/cmd_egg.c]: `} else {` .. dont put newlines here or just use r2-indent which does that for you

File: libr/core/cmd_egg.c [hunk 13]
```diff
@@ -107,13 +132,26 @@ static int cmd_egg(void *data, const char *input) {
...
+		else {
+			//nice error message ?
```
Comment (radare) [inline libr/core/cmd_egg.c]: yes please, or remove the empty block

**Comment by radare** [comment]
do not *ever* git pull in a branch you are exposing in a pullreq

**Comment by radare** [comment]
Fixed and merged

**Pull request closed without merging**